spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/LazyCurrentTraceContext.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.reactor;
+
+import brave.propagation.CurrentTraceContext;
+import brave.propagation.TraceContext;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.util.annotation.Nullable;
+
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.context.ConfigurableApplicationContext;
+
+/**
+ * Spaghetti jar for various hacks around bean factory, mostly to avoid calling the
+ * expensive {@link BeanFactory#getBean(Class)} many times.
+ *
+ * <p>
+ * This should be used in code that must use {@link CurrentTraceContext}, but is wired
+ * prior to context refresh.
+ *
+ * <p>
+ * This defers calls to {@link ConfigurableApplicationContext#getBean(Class)} to
+ * accommodate not yet {@link ConfigurableApplicationContext#refresh() refreshed}
+ * contexts. It saves the result to avoid performance penalties.
+ */
+final class LazyCurrentTraceContext extends CurrentTraceContext {
+
+	private static final Log log = LogFactory.getLog(LazyCurrentTraceContext.class);
+
+	final BeanFactory beanFactory;
+
+	final boolean instanceofConfigurableApplicationContext;
+
+	CurrentTraceContext currentTraceContext;
+
+	LazyCurrentTraceContext(BeanFactory beanFactory) {
+		this.beanFactory = beanFactory;
+		// There is an assumption that instanceof is limiting performance, though not
+		// verified here.
+		// https://github.com/reactor/reactor-core/pull/1882
+		this.instanceofConfigurableApplicationContext = beanFactory instanceof ConfigurableApplicationContext;
```

Comment:
I kept this special case though I think it is over optimization. I don't think we've done enough work on this code to start scrubbing like this.

Reply:
looks like you are right, it is over optimization as it is not a hot path